File: leetcode/solutions/0437.PathSumIII/pathSum.go
package solution

import (
	"github.com/yukisato/leetcode-util/datastructure"
)

// TreeNode structure for LeetCode
type TreeNode = datastructure.TreeNode

func pathSum(root *TreeNode, sum int) int {
	return dfs(root, sum, []int{})
}

func dfs(node *TreeNode, sum int, nums []int) int {
	if node == nil {
		return 0
	}

	count := 0
	nums = append(nums, node.Val)
	currentSum := 0

	for i := len(nums) - 1; i >= 0; i-- {
		currentSum += nums[i]

		if currentSum == sum {
			count++
		}
	}

	childrenTotalCount := dfs(node.Left, sum, nums) + dfs(node.Right, sum, nums)
	nums = nums[:len(nums)-1]

	return childrenTotalCount + count
}

File: leetcode/solutions/0437.PathSumIII/pathSum.ts
import { BinaryTreeNode } from "../../../datastructure/BinaryTreeNode.ts";

export default function pathSum(
  root: BinaryTreeNode<number> | null,
  sum: number
): number {
  const nums: number[] = [];
  let count = 0;

  function dfs(node: BinaryTreeNode<number> | null) {
    if (!node) {
      return;
    }

    nums.push(node.val);
    let check = 0;

    for (let i = nums.length - 1; i >= 0; i--) {
      check += nums[i];

      if (check === sum) {
        count++;
      }
    }

    dfs(node.left);
    dfs(node.right);

    nums.pop();
  }

  dfs(root);

  return count;
}
